nmostesting/mocks/Auth.py

```diff
@@ -0,0 +1,302 @@
+        self.protocol = "http"
+        self.host = get_default_ip()
+        if ENABLE_HTTPS:
+            self.protocol = "https"
+            self.host = get_mocks_hostname()
+
+    def make_mdns_info(self, priority=0, api_ver=None, ip=None):
+        """Get an mDNS ServiceInfo object in order to create an advertisement"""
+        if api_ver is None:
+            api_ver = self.version
+
+        api_proto = self.protocol
+
+        if ip is None:
+            ip = get_default_ip()
+            hostname = "nmos-mocks.local."
+        else:
+            hostname = ip.replace(".", "-") + ".local."
+
+        port = self.port
+
+        # TODO: Add another test which checks support for parsing CSV string in api_ver
+        txt = {'api_ver': api_ver, 'api_proto': api_proto, 'pri': str(priority)}
+
+        service_type = "_nmos-auth._tcp.local."
+        info = ServiceInfo(service_type,
+                           "NMOSTestSuite{}{}.{}".format(port, api_proto, service_type),
+                           addresses=[socket.inet_aton(ip)], port=port,
+                           properties=txt, server=hostname)
+        return info
+
+    def make_issuer(self):
+        return "{}://{}:{}".format(self.protocol, self.host, self.port)
+
+    def make_metadata(self):
+        metadata = {
+            "issuer": self.make_issuer(),
+            "authorization_endpoint": "{}://{}:{}/auth".format(self.protocol, self.host, self.port),
+            "token_endpoint": "{}://{}:{}/token".format(self.protocol, self.host, self.port),
+            "jwks_uri": "{}://{}:{}/jwks".format(self.protocol, self.host, self.port),
+            "registration_endpoint": "{}://{}:{}/register".format(self.protocol, self.host, self.port),
+            "response_types_supported": [
+                "code",
+                "code token"
+            ]
+        }
+        metadata["grant_types_supported"] = [e.name for e in GRANT_TYPES]
+        metadata["scopes_supported"] = [e.name for e in SCOPES]
+        metadata["code_challenge_methods_supported"] = [e.name for e in CODE_CHALLENGE_METHODS]
+        return metadata
+
+    def generate_jwk(self):
+        private_key = read_RSA_private_key(self.private_keys)
+        return generate_jwk(private_key)
+
+    def generate_token(self, scopes=None, write=False, azp=False, add_claims=True, overrides=None):
+        private_key = read_RSA_private_key(self.private_keys)
+        overrides_ = {"iss": "{}://{}:{}".format(self.protocol, self.host, self.port)}
+        if overrides:
+            overrides_.update(overrides)
+        return generate_token(private_key, scopes, write, azp, add_claims, overrides=overrides_)
+
+
+# 0 = Primary mock Authorization API
+# 1 = Secondary mock Authorization API, which is used for testing resources server on handling
+#     access token from different Authorization server
+NUM_AUTHS = 2
+AUTHS = [Auth(i + 1) for i in range(NUM_AUTHS)]
+AUTH_API = Blueprint('auth_pi', __name__)
```

Comment:
Unfortunately I don't feel qualified to review the Auth API implementation for correctness or completeness (sufficient to support most nodes/controllers under test)... @jonathan-r-thorpe, @cristian-recoseanu, is this something you can help with?

Reply:
In discussion today, we've decided that ideally the mock Auth server should validate requests with the schemas from the IS-10 repo and reject them as appropriate. Valid requests are required in order to subsequently pass the authorized API test cases, so although we won't with this PR report invalid requests on the Auth server via the test results, this is sufficient for now.

Reply:
It has now been done with schema validation on the `client registration` endpoint as this is the only endpoint expecting JSON request body and on the `client_assertion` (JWT). Extensive tests are added to the endpoint query parameters and the authorization header.

Reply:
To avoid doubt, you mean what the [RAML](https://github.com/AMWA-TV/is-10/blob/v1.0.x/APIs/AuthorizationAPI.raml) describes as `/{registration_endpoint_path}` and `/{token_endpoint_path}`, right?

Can you explain a bit about the fact that IS-10 says for both those endpoints the media type `application/x-www-form-urlencoded` should be used? The client registration request seems to be covered by [RFC 7591 Section 3.1](https://www.rfc-editor.org/rfc/rfc7591#section-3.1) which contains a(n admittedly non-normative) example using `application/json`... The token endpoint request is covered by [RFC 6749 Section 4.1.1](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.1) which does specify `application/x-www-form-urlencoded`.

Reply:
The `client registration` endpoint I mentioned is the RAML `/{registration_endpoint_path}`. The client details are as the JSON payload.

https://www.rfc-editor.org/rfc/rfc7591#section-3.1
> To register, the client or developer sends an HTTP POST to the client
 registration endpoint with a content type of `"application/json"`.  The
 HTTP Entity Payload is a JSON [[RFC7159](https://www.rfc-editor.org/rfc/rfc7159)] document consisting of a JSON
 object and all requested client metadata values as top-level members
 of that JSON object.

The AThe `/token` endpoint using media type application/x-www-form-urlencoded

Reply:
> The `client registration` endpoint I mentioned is the RAML `/{registration_endpoint_path}`. The client details are as the JSON payload.

So IS-10 is wrong...

Reply:
> The [RFC 6749 Section 4.1.1](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.1) is talking about the `Authorization Request`, i.e. the `/auth` endpoint, not the `/register`.

You're right, I incorrectly said 4.1.1 when I meant [RFC 6749 Section 4.1.3](https://www.rfc-editor.org/rfc/rfc6749#section-4.1.3), as the spec for the `/{token_endpoint_path}`, sorry!

Reply:
> > The `client registration` endpoint I mentioned is the RAML `/{registration_endpoint_path}`. The client details are as the JSON payload.
> 
> So IS-10 is wrong... specifically this:
> 
> https://github.com/AMWA-TV/is-10/blob/518b4532225ff5503169903409c98db6c8bd294b/APIs/AuthorizationAPI.raml#L166-L167

I think the IS-10 spec is wrong. Registration data do not pass in as `application/x-www-form-urlencoded`

Reply:
https://github.com/AMWA-TV/is-10/issues/113